Implement MIDI input support

HoLLy-HaCKeR opened the pull request:
Closes #2723.

Still WIP, uses [managed-midi](https://github.com/atsushieno/managed-midi) for cross-platform MIDI input.

Remaining issues:
- [x] ~~[Running status](https://web.archive.org/web/20130305092440/http://home.roadrunner.com/~jgglatt/tech/midispec/run.htm) is not handled yet, which is used by Linux when multiple keys are pressed at once.~~
- [x] ~~Linux midi input stops when too many keys are pressed at once~~
- [x] ~~Disconnecting a device on Windows throws an unhandled exception (atsushieno/managed-midi#49)~~
- [x] ~~Does not yet take velocity into account~~

Tested on:
- [x] Windows
- [x] MacOS
- [x] Linux
- [ ] Android
- [ ] iOS

iiSaLMaN commented on osu.Framework/Input/Handlers/MidiInputHandler.cs:
```diff
@@ -0,0 +1,107 @@
+        public override bool Initialize(GameHost host)
+        {
+            // Try to initialize. This can throw on Linux if asound cannot be found.
+            try {
```
Keep in mind we have open block in a new line
```suggestion
            try
            {

HoLLy-HaCKeR replied:
I know, this PR is not yet done, I'll fix all AppVeyor and CodeFactor issues when it is. It's currently open for feedback related to code structure and implementation details, and visibility.

iiSaLMaN replied:
> I know, this PR is not yet done, I'll fix all AppVeyor and CodeFactor issues when it is. It's currently open for feedback related to code structure and implementation details, and visibility.

Otherwise, this change is fantastic, great job on this one 👍

iiSaLMaN commented on osu.Framework/Input/MidiKey.cs:
```diff
@@ -0,0 +1,105 @@
+{
+    public enum MidiKey
+    {
+        A0 = 21,
```
where did this number come from?

HoLLy-HaCKeR replied:
Each key on a piano has a corresponding MIDI number. A0 happens to be 21. See more [here](https://newt.phys.unsw.edu.au/jw/notes.html).

HoLLy-HaCKeR commented:
Point raised in private discussion: Should MidiKey be part of InputKey? Right now it's a separate enum which makes conversion to and from integers easy, but it probably isn't usable as an input in projects like lazer.

peppy commented:
It should be available as normal input, yes.

HoLLy-HaCKeR commented:
That CodeFactor failure complains about `KeyBindingContainer.Handle(UIEvent)` being too complex, but it contains only a switch statement so I don't know if it's logical to split that up into different methods.

HoLLy-HaCKeR commented:
PR should be ready for review and testing on Android, iOS and MacOS. I've tried testing on Android but my device doesn't seem to support MIDI input.

smoogipoo commented:
Unfortunately this library seems to crash on macOS (https://github.com/atsushieno/managed-midi/issues/50), and the developer seems to be "unable to continue development" (https://github.com/atsushieno/managed-midi/issues/42).

May be best to wait for ManagedBass to be fixed...

Code-wide it looks great.

smoogipoo commented:
Have made a few fixes for other hard crashes in managed-midi, but at this point I'm not completely sold on the stability of the framework. Gonna delay merging this one for a few weeks while the managedbass guy responds (they're generally pretty responsive).

HoLLy-HaCKeR commented:
@smoogipoo ManagedBass got updated to v3 and I got a response in ManagedBass/ManagedBass#54, and it seems Bass does not support MIDI input on Android. Managed-midi has received a few bugfixes since your last message so could you give it another look when you have a spare moment?

smoogipoo commented:
I'm able to trigger the sequence `Event F0:7E:7F` by refreshing devices on macOS. Based on how MIDI input is implemented on macOS it's probably doing something really silly, but this should probably be guarded against otherwise it crashes the app. Causes:

```
Unhandled exception. [runtime] 2020-04-27 04:45:45 [error]: An unhandled error has occurred.
[runtime] 2020-04-27 04:45:45 [error]: System.Collections.Generic.KeyNotFoundException: The given key '__Source0' was not present in the dictionary.
[runtime] 2020-04-27 04:45:45 [error]: at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
[runtime] 2020-04-27 04:45:45 [error]: at osu.Framework.Input.Handlers.Midi.MidiInputHandler.readEvent(Byte[] data, String senderId, Int32& i, Byte& eventType, Byte& key, Byte& velocity) in /Users/smoogipoo/Repos/osu-framework/osu.Framework/Input/Handlers/Midi/MidiInputHandler.cs:line 120
[runtime] 2020-04-27 04:45:45 [error]: at osu.Framework.Input.Handlers.Midi.MidiInputHandler.onMidiMessageReceived(Object sender, MidiReceivedEventArgs e) in /Users/smoogipoo/Repos/osu-framework/osu.Framework/Input/Handlers/Midi/MidiInputHandler.cs:line 108
[runtime] 2020-04-27 04:45:45 [error]: at Commons.Music.Midi.CoreMidiApi.CoreMidiInput.OnMessageReceived(Object sender, MidiPacketsEventArgs e)
[runtime] 2020-04-27 04:45:45 [error]: at CoreMidi.MidiPort.CallMessageReceived(IntPtr pktlist, IntPtr readProcRefCon, IntPtr srcConnRefCon)
```

smoogipoo commented on osu.Framework/Input/InputManager.cs:
```diff
@@ -539,6 +560,10 @@ protected virtual void HandleMouseButtonStateChange(ButtonStateChangeEvent<Mouse
 
+        private bool handleMidiKeyDown(InputState state, MidiKey key, byte velocity) => PropagateBlockableEvent(NonPositionalInputQueue, new MidiDownEvent(state, key, velocity));
+
+        private bool handleMidiKeyUp(InputState state, MidiKey key) => PropagateBlockableEvent(NonPositionalInputQueue, new MidiUpEvent(state, key));
```
This should be updated to use the new `ButtonEventManager` stuff. See how it's implemented for joysticks / mouse buttons / keys for reference.

smoogipoo commented on osu.Framework/Graphics/Drawable.cs:
```diff
@@ -2192,6 +2198,10 @@ public bool TriggerEvent(UIEvent e)
 
+        protected virtual bool OnMidiDown(MidiDownEvent e) => Handle(e);
+
+        protected virtual bool OnMidiUp(MidiUpEvent e) => Handle(e);
```
Should return `void` (and `false` in `TriggerEvent()`).

smoogipoo commented on osu.Framework/osu.Framework.csproj:
```diff
@@ -21,6 +21,7 @@
     <AllowedOutputExtensionsInPackageBuildOutputFolder>$(AllowedOutputExtensionsInPackageBuildOutputFolder);.pdb</AllowedOutputExtensionsInPackageBuildOutputFolder>
   </PropertyGroup>
   <ItemGroup Label="Package References">
+    <PackageReference Include="managed-midi" Version="1.9.8" />
```
Update this while you're here.

smoogipoo commented:
Also I assume this isn't supposed to support channels at this point? It's coded in such a way that it seems like it should support them, but it isn't for some reason.

atsushieno commented:
In case you are trying to use managed-midi's own Core MIDI support, note that its quality is super low and not quite ready for general consumption yet. I cannot provide solid implementation for Xamarin.Mac alternatives up to CoreFoundation level. On Mac managed-midi **via Xamarin.Mac** is the only viable approach (managed-midi has particular build for the profile). Or you might want to try RtMidiSharp which might work better.

smoogipoo commented:
Ah... We're avoiding using Xamarin on desktop platforms in favour of netcore, so that isn't really an option for us at this point. Maybe NET5 will change that.

I'm not super concerned about macOS support for the time being though, just as long as it doesn't crash the app. If the reason for channels not working is a result of being on macOS then I'm fine with that.

atsushieno commented:
Sorry but I'm not really tracking this project issue and have no idea what kind of issue the "channels" stuff is. If there is something reproducible at my side let me know and I might be able to help.

HoLLy-HaCKeR commented:
Updated according to feedback, there's a few things I want to mention:
- `MidiKeyInput` does not extend `ButtonInput<TKey>`, but implements `IInput` directly. `ButtonInput<TKey>`'s purpose seems to be to handle multiple input changes at once, but `MidiKeyInput` is only directly used in `MidiInputHandler` so it doesn't seem useful.
- I changed how velocity gets stored. `MidiState` now stores an internal `Dictionary<MidiKey, byte>`, which gets updated in `MidiStateChangeEvent`'s ctor and read in `MidiKeyEventManager.HandleButtonDown`. It's not the prettiest solution, but the only other way I could think of involved refactoring `ButtonEventManager` which seems a bit silly if it's only for MIDI input. I can try to look for something better if this isn't good enough.
- MIDI input is currently accepted even when the game is not focused, though most other programs share this behavior.

smoogipoo commented:
The way you've implemented it is fine :) Made one minor change to make the velocities public - may come in useful if someone's handling input manually via `InputManager.CurrentState`.

Thanks!
